### PR TITLE
OpenVASP Register

### PR DESCRIPTION
**Pull request by DanielSollis**
Completes Implementation of the openvasp register endpoint.

### CODE REVIEW

File: pkg/openvasp/web-service-gin/openvasp.go [hunk 1]
```diff
@@ -30,41 +35,86 @@ const (
...
+
+func Serve(address, dns string) (err error) {
+	if db, err = OpenDB(dns); err != nil {
+		return err
+	}
+
+	router := gin.Default()
+	router.POST("/register", Register)
+	router.POST("/transfer", Transfer)
+	router.Run(fmt.Sprintf("localhost%s", address))
```
Comment (pdeziel) [inline pkg/openvasp/web-service-gin/openvasp.go:50]: I'm a bit confused why this is necessary, it looks like the CLI is already passing in the full addrport?

**Review by pdeziel — COMMENTED** [review]
Looks like a good start, I did have some suggestions to make this a bit more user-friendly and also help our future selves a bit.

I assume you're just using curl to test this right now? Would it be helpful to have client commands in the CLI or actual unit tests?

File: pkg/openvasp/web-service-gin/openvasp.go [hunk 2]
```diff
@@ -2,18 +2,23 @@ package openvasp
...
 )
 
 type Customer struct {
-	ID            uuid.UUID
-	Name          string
-	AssetType     VirtualAsset
-	WalletAddress string
-	TravelAddress string
+	gorm.Model
+	CustomerID    uuid.UUID    `gorm:"uniqueIndex;size:255;column:customer_id;not null"`
```
[the remark this answers is not in the document]
Comment (DanielSollis) [reply pkg/openvasp/web-service-gin/openvasp.go:17]: Thanks!

File: pkg/openvasp/web-service-gin/openvasp.go [hunk 3]
```diff
@@ -30,41 +35,100 @@ const (
...
+		return nil, err
+	}
+
+	if err = db.AutoMigrate(&Customer{}); err != nil {
+		return nil, err
+	}
+	return db, nil
+}
+
+func Serve(address, dsn string) (err error) {
```
Comment (pdeziel) [inline pkg/openvasp/web-service-gin/openvasp.go:63]: Much better!

File: pkg/openvasp/web-service-gin/openvasp.go [hunk 4]
```diff
@@ -30,41 +35,100 @@ const (
...
+			--include \
+			--header "Content-Type: application/json" \
+			--request "POST" --data '{"name":"Tildred Milcot", "assettype": 3, "walletaddress": "926ca69a-6c22-42e6-9105-11ab5de1237b"}'
+*/
+func (s *server) Register(c *gin.Context) {
 	var err error
-	if err = c.BindJSON(newCustomer); err != nil {
+	var newCustomer Customer
+	if err = c.BindJSON(&newCustomer); err != nil {
+		c.IndentedJSON(http.StatusBadRequest, gin.H{"JSON Binding error": err.Error()})
```
Comment (pdeziel) [inline pkg/openvasp/web-service-gin/openvasp.go:91]:
```suggestion
		c.IndentedJSON(http.StatusBadRequest, gin.H{"Could not bind request": err.Error()})
```

File: pkg/openvasp/web-service-gin/openvasp.go [hunk 5]
```diff
@@ -30,41 +35,100 @@ const (
...
+		return
+	}
+
+	if err = validateCustomer(&newCustomer); err != nil {
+		c.IndentedJSON(http.StatusBadRequest, gin.H{"validation error": err.Error()})
```
Comment (pdeziel) [inline pkg/openvasp/web-service-gin/openvasp.go:96]:
```suggestion
		c.IndentedJSON(http.StatusBadRequest, gin.H{"Invalid customer provided": err.Error()})
```

File: pkg/openvasp/web-service-gin/openvasp.go [hunk 6]
```diff
@@ -30,41 +35,100 @@ const (
...
 		return
 	}
 
-	if err = validateCustomer(newCustomer); err != nil {
+	travelAddress := fmt.Sprintf(travelURLTemplate, newCustomer.WalletAddress)
+	if newCustomer.TravelAddress, _, err = lnurl.GenerateLnUrl(travelAddress); err != nil {
+		c.IndentedJSON(http.StatusBadRequest, gin.H{"LNURL generation error": err.Error()})
```
Comment (pdeziel) [inline pkg/openvasp/web-service-gin/openvasp.go:102]:
```suggestion
		c.IndentedJSON(http.StatusInternalServerError, gin.H{"Could not register customer": err.Error()})
```

File: pkg/openvasp/web-service-gin/openvasp.go [hunk 7]
```diff
@@ -30,41 +35,100 @@ const (
...
 		return
 	}
 
-	//TODO: generate LNURL TravelAddress
+	if db := s.db.Create(&newCustomer); db.Error != nil {
+		c.IndentedJSON(http.StatusInternalServerError, gin.H{"Database error": db.Error})
```
Comment (pdeziel) [inline pkg/openvasp/web-service-gin/openvasp.go:107]:
```suggestion
		c.IndentedJSON(http.StatusInternalServerError, gin.H{"Could not register customer": db.Error})
```

**Review by pdeziel — APPROVED** [review]
Thanks for making the changes, I've included some suggestions for the error codes and messages so they make more sense to users.

**Pull request merged**